lambda-functions/cleanup-unused-images/main.go

```diff
@@ -0,0 +1,163 @@
+		Region: aws.String(os.Getenv("AWS_REGION"))},
+	)
+	if err != nil {
+		log.WithError(err).Error("AWS Session failed.")
+		return
+	}
+	svc := ec2.New(sess)
+	uniqueUsedImages, err := getUniqueUsedImages(svc)
+	err = deleteAMIs(svc, uniqueUsedImages)
+
+	if err != nil {
+		log.WithError(err).Error("Failed to delete AMIs.")
+		return
+	}
+}
+
+func deleteAMIs(svc *ec2.EC2, uniqueUsedImages []string) error {
+	imagesInput := &ec2.DescribeImagesInput{
+		Owners: []*string{
+			aws.String(os.Getenv("OWNER_ID")),
+		},
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String("mattermost-cloud-*")},
+			},
+		},
+	}
+	snapshots, err := getAllSnapshots(os.Getenv("OWNER_ID"), svc)
+	if err != nil {
+		log.WithError(err).Error("Failed to get Snapshots.")
+		return err
+
+	}
+	allImages, err := svc.DescribeImages(imagesInput)
+	if err != nil {
+		log.WithError(err).Error("Failed to describe images.")
+		return err
+
+	}
+	oldImages, err := filterImagesByDateRange(allImages.Images, 730)
+	if err != nil {
+		log.WithError(err).Error("Failed to filter images by date range.")
+		return err
+
+	}
+	dryRun := true
+	for _, i := range oldImages {
+		imageForCleanup := contains(uniqueUsedImages, *i.ImageId)
+		fmt.Println(*i.ImageId + ": De-registering AMI named \"" + *i.Name + "\"...")
+		cleanupImageInput := &ec2.DeregisterImageInput{
+			ImageId: &imageForCleanup,
+			DryRun:  &dryRun,
+		}
+		_, err := svc.DeregisterImage(cleanupImageInput)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to deregister AMI %s", *i.ImageId)
+		}
+		var snapshotIds []string
+		for _, snapshot := range snapshots {
+			if strings.Contains(*snapshot.Description, *i.ImageId) {
+				snapshotIds = append(snapshotIds, *snapshot.SnapshotId)
+			}
+		}
+		fmt.Println(*i.ImageId + ": Found " + strconv.Itoa(len(snapshotIds)) + " snapshot(s) to delete")
+		for _, snapshotId := range snapshotIds {
+			fmt.Println(*i.ImageId + ": Deleting snapshot " + snapshotId + "...")
+			_, deleteErr := svc.DeleteSnapshot(&ec2.DeleteSnapshotInput{
+				DryRun:     &dryRun,
+				SnapshotId: &snapshotId,
+			})
+
+			if deleteErr != nil {
+				log.WithError(deleteErr).Errorf("Failed to delete Snapshot %s", snapshotId)
+			}
+		}
+
+	}
+	return nil
+}
+
+func getUniqueUsedImages(svc *ec2.EC2) ([]string, error) {
+
+	instancesInput := &ec2.DescribeInstancesInput{}
+	encountered := make(map[string]bool)
+	runningInstances, err := svc.DescribeInstances(instancesInput)
+	if err != nil {
+		return nil, err
+	}
+	for _, i := range runningInstances.Reservations {
+		for _, k := range i.Instances {
+			encountered[*k.ImageId] = true
+		}
+	}
+
+	uniqueUsedImages := []string{}
+	for image := range encountered {
+		uniqueUsedImages = append(uniqueUsedImages, image)
+	}
+
+	return uniqueUsedImages, nil
+}
+
+func contains(arr []string, str string) string {
+	for _, a := range arr {
+		if a == str {
+			continue
+		}
+	}
+	return str
+}
```

Comment:
Hmm, I am not sure that this accomplishes what you are looking for. This will always return the string that is passed into it regardless of what the slice is.

Are you looking for something like this instead?

```
func contains(arr []string, str string) bool {
    for _, a := range arr {
        if a == str {
            return true
        }
    }
    return false
}
```

Reply:
I tested it on my local and getting the expected results.

Reply:
Hmm, I am a bit worried because it looks like this code will return the string you pass in every time you call the function. Let's get another reviewers eyes on it to make sure I am not missing something.

Reply:
I am confused on that as well tbh. In this function you want to check if a list has a specific entry and return this entry right? If that's the case I think what Gabe suggested offers the solution.  The function you are using is basically returning the string that it needs to check if it belongs to the slice at all times.

Reply:
@gabrieljackson 's suggestion doesn't do what I need here. I want to return the string itself when is not contained to the slice.

Reply:
@stafot I am still confused tbh. Lets examine a case. If the string is contained in the slice what will it do? It will go to continue, which means it will move to the next iteration of the loop. And when it finishes all iterations it will return the string. So it will return the string no matter if it exists in the slice or not. Makes sense?

Reply:
@stylianosrigas No, this is not the case in my understanding. It returns only the strings that are not contained to the uniqueUsedImages. Which is the desired functionality.

Reply:
@stafot But the code says otherwise. The function `contains` returns only a variable called str which is the variable passed in the function. The function gets an input for example str = ami-xxx and will return ami-xxx no matter what happens. Variable `str` will return no matter if it exists in the slice or not.

Reply:
Exactly, no matter which string and slice/array is passed into `contains` it will always return the string. It doesn't matter if the slice has the string or not. Right now the function isn't actually checking anything. The result will always be the same.

Reply:
Yes you are both right. Thank you both @gabrieljackson and @stylianosrigas  for pointing it. The results where the expected because the filtering by date were removing the used images.

Reply:
No problem, @stafot. Glad we are on the same page. I just wanted to be careful when doing the cleanup. Let us know if we can help at all with any of the changes that you will be making.